crates/core/tedge_api/src/entity.rs

```diff
@@ -0,0 +1,168 @@
+#[serde(transparent)]
+pub struct EntityExternalId(String);
+
+impl AsRef<str> for EntityExternalId {
+    fn as_ref(&self) -> &str {
+        &self.0
+    }
+}
+
+// XXX: As `EntityExternalId` is used as a part of cloudbound MQTT topic, it
+// can't contain characters invalid in topics, i.e. `+` and `#`. ([MQTT-4.7]).
+// If it's derived from a MQTT topic, this holds, but if created from a string,
+// this isn't checked, which is invalid!
+impl From<&str> for EntityExternalId {
+    fn from(val: &str) -> Self {
+        Self(val.to_string())
+    }
+}
+
+impl From<&String> for EntityExternalId {
+    fn from(val: &String) -> Self {
+        Self(val.to_string())
+    }
+}
+
+impl From<String> for EntityExternalId {
+    fn from(val: String) -> Self {
+        Self(val)
+    }
+}
+
+impl From<EntityExternalId> for String {
+    fn from(value: EntityExternalId) -> Self {
+        value.0
+    }
+}
+
+impl From<&EntityExternalId> for String {
+    fn from(value: &EntityExternalId) -> Self {
+        value.0.clone()
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct EntityMetadata {
+    #[serde(rename = "@topic-id")]
+    pub topic_id: EntityTopicId,
+    #[serde(rename = "@parent")]
+    pub parent: Option<EntityTopicId>,
+    #[serde(rename = "@type")]
+    pub r#type: EntityType,
+    #[serde(rename = "@id", skip_serializing_if = "Option::is_none")]
+    pub external_id: Option<EntityExternalId>,
+
+    // TODO: use a dedicated struct for cloud-specific fields, have `EntityMetadata` be generic over
+    // cloud we're currently connected to
+    #[serde(flatten)]
+    pub other: Map<String, JsonValue>,
+    #[serde(skip)]
+    pub twin_data: Map<String, JsonValue>,
+}
+
+impl EntityMetadata {
+    /// Creates a entity metadata for the main device.
+    pub fn main_device(device_id: String) -> Self {
+        Self {
+            topic_id: EntityTopicId::default_main_device(),
+            external_id: Some(device_id.clone().into()),
+            r#type: EntityType::MainDevice,
+            parent: None,
+            other: Map::new(),
+            twin_data: Map::new(),
+        }
+    }
+
+    /// Creates a entity metadata for a child device.
+    pub fn child_device(child_device_id: String) -> Result<Self, TopicIdError> {
+        Ok(Self {
+            topic_id: EntityTopicId::default_child_device(&child_device_id)?,
+            external_id: Some(child_device_id.clone().into()),
+            r#type: EntityType::ChildDevice,
+            parent: Some(EntityTopicId::default_main_device()),
+            other: Map::new(),
+            twin_data: Map::new(),
+        })
+    }
+
+    pub fn display_name(&self) -> Option<&str> {
+        self.other.get("name").and_then(|v| v.as_str())
+    }
+
+    pub fn display_type(&self) -> Option<&str> {
+        self.other.get("type").and_then(|v| v.as_str())
+    }
+
+    pub fn external_id_unchecked(&self) -> &EntityExternalId {
```

Comment:
This function was one side effect of reusing the same `EntityMetadata` struct in both entity store and entity cache. When used in the entity store, this value may or may not be present (based on the presence of `@id` in the registration message). But, when used in the entity cache, the external id is guaranteed to be there (auto-generated when not provided via `@id` during registration). So, the mapper code always uses this function with that implicit guarantee, to directly retrieve the external id.

Reply:
See https://github.com/thin-edge/thin-edge.io/pull/3230#discussion_r1910506459 for a proposal to fix that.